File: src/main/linkedlist/python/linked_list.py
class linked_list:
    """
    Representação de uma lista encadeada
    """
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0


    def is_empty(self):
        """
        Verifica se a lista está vazia

        Returns:
                boolean: true, caso a lista esteja vazia, false caso não
        """
        return self.size == 0


    def add_first(self, value):
        """
        Adiciona um elemento na primeira posiçao da lista

        Args:
            value (int): o valor a ser adicionado
        """
        new_node = node(value, self.head)
        
        if self.size == 0:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node
        
        self.size += 1


    def add_last(self, value):
        """
        Adiciona um elemento na última posiçao da lista

        Args:
            value (int): o valor a ser adicionado
        """
        new_node = node(value)

        if self.size == 0:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        
        self.size += 1


    def add(self, value, index):
        """
        Adiciona um elemento em um índice específico da lista

        Args:
            value (int): o valor a ser adicionado
            index (int): a posição na qual ele será adicionado
        """
        if index == 0:
            self.add_first(value)
        elif (index == self.size)
            self.add_last(value)
        else:
            if index < 0 or index > self.size-1: raise IndexError("Índice inválido")

            new_node = node(value)
            prev_node = self.__get_node_by_index(index-1)

            new_node.next = prev_node.next
            prev_node.next = new_node

            self.size += 1

    
    def get_first(self):
        """
        Informa o valor do primeiro elemento da lista

        Returns:
                int: o valor do primeiro elemento da lista
        """
        return self.head.value


    def get_last(self):
        """
        Informa o valor do primeiro elemento da lista

        Returns:
                int: o valor do último elemento da lista
        """
        return self.tail.value


    def get(self, index):
        """
        Informa o valor do node a partir de seu índice

        Args:
            index (int): o índice do node
        Returns:
                int: o valor contido no node
        """
        if index < 0 or index > self.size-1: raise IndexError("Índice inválido")

        if index == 0:
            return self.get_first()
        elif index == self.size-1:
            return self.get_last()
        
        return self.__get_node_by_index(index).value


    def update_node(self, value, index):
        """
        Altera o valor de um nó na lista

        Args:
            value (int): o valor a ser alterado
            index (int): o índice do valor a ser alterado
        """
        if index < 0 or index > self.size-1: raise IndexError("Índice inválido")

        node = self.__get_node_by_index(index)
        node.value = value


    def remove_first(self):
        """Remove o primeiro elemento da lista"""
        if self.is_empty(): raise Exception("Lista vazia")
        
        if self.size == 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next

        self.size -= 1


    def remove_last(self):
        """Remove o último elemento da lista"""
        if self.is_empty(): raise Exception("Lista vazia")
        
        if self.size == 1:
            self.head = None
            self.tail = None
        else:
            prev_node = self.__get_node_by_index(self.size-2)
            prev_node.next = None
            self.tail = prev_node
        
        self.size -= 1
    

    def remove_by_index(self, index):
        """
        Remove um elemento da lista a partir de seu índice

        Args:
            index (int): o índice do elemento a ser removido
        """
        if index < 0 or index > self.size-1: raise IndexError("Índice inválido")

        if index == 0:
            self.remove_first()
        elif index == self.size-1:
            self.remove_last()
        else:
            prev_node= self.__get_node_by_index(index-1)
            prev_node.next = prev_node.next.next

            self.size -= 1


    def remove_by_value(self, value):
        """
        Remove o primeiro elemento da lista que tiver o valor informado

        Args:
            value (int): o valor a ser removido
        """
        index = self.index_of(value)

        if index != -1:
            self.remove_by_index(index)    


    def index_of(self, value):
        """
        Informa o índice da primeira ocorrência de um valor. Se não estiver presente na lista, retornará -1

        Args:
            value (int): o valor a ser encontrado

        Returns:
            int: o índice da primeira ocorrência do valor, caso esteja na lista, -1 caso não.
        """
        if self.is_empty(): raise Exception("Lista vazia")

        node = self.head
        for i in range(self.size):
            if node.value == value:
                return i
        
            node = node.next
        
        return -1


    def last_index_of(self, value):
        """
        Informa o índice da última ocorrência de um valor. Se não estiver presente na lista, retornará -1

        Args:
            value (int): o valor buscado

        Returns:
            int: o índice da última ocorrência do valor, caso esteja na lista, -1 caso não
        """
        if self.is_empty(): return

        index = -1

        node = self.head
        for i in range(self.size):
            if node.value == value:
                index = i
            
            node = node.next
        
        return index


    def __get_node_by_index(self, index):
        """
        Pega o node em um índice informado

        Args:
            index (int): o índice do node
        
        Returns:
            node: o node
        """
        if index < 0 or index > self.size-1: return

        node = self.head
        for i in range(index):
            node = node.next
        
        return node


    def size_ll(self):
        """
        Informa o tamanho da lista

        Returns:
                int: o tamanho da lista
        """
        return self.size


    def contains(self, value):
        """
        Verifica se um valor está contido ou não na lista
        
        Args:
            value (int): o valor a ser buscado

        Returns:
                boolean: true caso o valor esteja contido, false caso não
        """
        return self.index_of(value) != -1


    def toString(self):
        """Representação textual da lista no formato "a -> b -> c [...]"""
        llString = ""   
        node = self.head
        for i in range(self.size):
            llString += node.value
            if node.next != None:
                llString += " -> "
            
            node = node.next
        
        return llString
    

class node:
    """Representação de um Nó da lista"""
    def __init__(self, value=None):
        self.value = value
        self.next = None

File: src/main/linkedlist/java/LinkedList.java
import java.util.NoSuchElementException;

/**
 * Representação de uma lista encadeada
 */
public class LinkedList {
    /**
     * O primeiro nó da lista
     */
    private Node head;

    /**
     * O último nó da lista
     */
    private Node tail;

    /**
     * O tamanho da lista
     */
    private int size;
    
    public LinkedList() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }
    
    /**
     * Verifica se a lista está vazia
     * @return true caso esteja vazia, false caso não
     */
    public boolean isEmpty() {
        return this.size == 0;
    }

    /**
     * Adiciona um elemento na primeira posição da lista
     * 
     * @param value o valor a ser adicionado
     */
    public void addFirst(int value) {
        Node newNode = new Node(value);

        if (isEmpty()) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.size++;
    }

    /**
     * Adiciona um elemento na última posição da lista
     * 
     * @param value o valor a ser adicionado
     */
    public void addLast(int value) {
        Node newNode = new Node(value);
        
        if (isEmpty()) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.size++;
    }

    /**
     * Adiciona um elemento em um índice específico da lista
     * 
     * @param value o valor a ser adicionado
     * @param index a posição na qual ele será adicionado
     */
    public void add(int value, int index) {        
        if (index == 0)
            addFirst(value);
        else if (index == this.size)
            addLast(value);
        else {
            if (index < 0 || index > this.size-1) throw new IndexOutOfBoundsException("Índice inválido");
            
            Node newNode = new Node(value);
            Node prevNode = getNodeByIndex(index-1);
            
            newNode.next = prevNode.next;
            prevNode.next = newNode;
            this.size++;
        }
    }

    /**
     * Informa o valor do primeiro elemento da lista
     * 
     * @return o valor do primeiro elemento da lista
     */
    public int getFirst() {
        return this.head.value;
    }

    /**
     * Informa o valor do último elemento da lista
     * 
     * @return o último elemento da lista
     */
    public int getLast() {
        return this.tail.value;
    }

    /**
     * Pega o valor do node a partir de seu índice
     * 
     * @param index o índice do node
     * @return o valor contido no node
     */
    public int get(int index) {
        if (index < 0 || index > this.size-1) throw new IndexOutOfBoundsException("Índice inválido");
        
        int value;
        if (index == 0)
            value = getFirst();
        else if (index == this.size-1)
            value = getLast();
        else
            value = getNodeByIndex(index).value;
            
        return value;
    }

    /**
     * Altera o valor de um nó na lista
     * 
     * @param value o valor a ser alterado
     * @param index o índice do valor a ser alterado
     */
    public void updateNode(int value, int index) {
        if (index < 0 || index > this.size-1) throw new IndexOutOfBoundsException("Índice inválido");

        Node node = getNodeByIndex(index);
        node.value = value;
    }

    /**
     * Remove o primeiro elemento da lista
     */
    public void removeFirst() {
        if (isEmpty()) throw new NoSuchElementException("Node não existe");

        if (this.size == 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.next;            
        }
        this.size--;
    }

    /**
     * Remove o último elemento da lista
     */
    public void removeLast() {
        if (isEmpty()) throw new NoSuchElementException("Node não existe");

        if (this.size == 1) {
            this.head = null;
            this.tail = null;
        } else {
            Node prevNode = getNodeByIndex(this.size-2);
            prevNode.next = null;
            this.tail = prevNode;
        }
        this.size--;
    }

    /**
     * Remove um elemento da lista a partir de seu índice
     * 
     * @param index o índice do elemento a ser removido
     */
    public void removeByIndex(int index) {
        if (index < 0 || index > this.size-1) throw new IndexOutOfBoundsException("Índice inválido");
        
        if (index == 0)
            removeFirst();
        else if (index == this.size-1)
            removeLast();
        else {
            Node prevNode = getNodeByIndex(index-1);
            prevNode.next = prevNode.next.next;

            this.size--;
        }
    }

    /**
     * Remove o primeiro elemento da lista que tiver o valor informado
     * 
     * @param value o valor a ser removido
     */
    public boolean removeByValue(int value) {        
        int index = indexOf(value);

        if (index != -1) {
            removeByIndex(index);
            return true;
        }
        return false;
    }   
    
    /**
     * Informa o índice da primeira ocorrência de um valor. Se não estiver presente na lista, retornará -1
     * 
     * @param value o valor a ser encontrado
     * @return o índice da primeira ocorrência do valor, caso esteja na lista, -1 caso não.
     */
    public int indexOf(int value) {
        if (isEmpty()) throw new NoSuchElementException("Elemento não existe");

        Node node = this.head;
        for (int i = 0; i < this.size; i++) {
            if (node.value == value)
                return i;

            node = node.next;
        }
        return -1;
    }

    /**
     * Informa o índice da última ocorrência de um valor. Se não estiver presente na lista, retornará -1
     * 
     * @param value o valor buscado
     * @return o índice da última ocorrência do valor, caso esteja na lista, -1 caso não
     */
    public int lastIndexOf(int value) {
        if (isEmpty()) throw new NoSuchElementException("Elemento não existe");
        
        int index = -1;
        
        Node node = this.head;
        for (int i = 0; i < this.size; i++) {
            if (node.value == value)
                index = i;
            
            node = node.next;
        }
        return index;
    }

    /**
     * Pega o node em um índice informado
     * 
     * @param index o índice do node
     * @return o node
     */
    private Node getNodeByIndex(int index) {
        if (index < 0 || index > this.size-1) throw new IndexOutOfBoundsException("Índice inválido");

        Node node = this.head;
        for (int i = 0; i < index; i++)
        node = node.next;
        
        return node;
    }

    /**
     * Informa o tamanho da lista
     * 
     * @return o tamanho da lista
     */
    public int sizeLL() {
        return this.size;
    }
    
    /**
     * Verifica se um valor está contido ou não na lista
     * 
     * @param value o valor a ser buscado
     * @return true caso o valor esteja contido, false caso não
     */
    public boolean contains(int value) {
       return indexOf(value) != -1;
    }

    /**
     * Representação textual da lista no formato "a -> b -> c [...]"
     */
    public String toString() {
        if (isEmpty()) return "";

        Node currNode = this.head;
        String llString = "";
        while (currNode != null) {
            llString += currNode.value + " -> ";
            currNode = currNode.next;
        }
        return llString.substring(0, llString.length() - 4);
        }
    
    /**
     * Representação de um Nó da lista
     */
    public static class Node {
        int value;
        Node next;
        
        public Node(int value) {
            this.value = value;
            this.next = null;
        }
    }
    
}
